### PR TITLE
Pass current state of the definition to definition_helpers in APISpec.definition

### PR DESCRIPTION
**Pull request by martinlatrille**
Now pass the current state of the definition (i.e. `ret`) as a `kwarg` to all definition helpers.

It will allow helpers to do modifications of nested value in the definition (like modifying a schema's property for example) without overriding the whole top-level dict (`properties`, in our example).

### CODE REVIEW

**Comment by lafrech** [comment]
Looks good.

Note that this relies on the order in which the helpers are executed, which is the registration order: the order of the plugins in the list you pass at APISpec init. This is also the case for path helpers. This could be an issue if you need a different order for path helpers and definition helpers. But this is more of a theoretical issue, I don't think it would be blocking in practice. Anyway, there's no reason to block this PR until this is addressed.

Care to add a test before we can merge this?

Also, I suggest this change:

```python
- Definition: the current state of the definition of the schema
+ definition (dict): current state of the definition
```

File: apispec/core.py
Comment (sloria) [inline apispec/core.py]: To piggy-back on @lafrech 's suggestion, how about:

```
``kwargs`` will include (among other things):
- definition (dict): current state of the definition
```

**Review by sloria — CHANGES_REQUESTED** [review]
Once the test and small doc change have been made, this is good to merge. Thanks for the PR, and thanks @lafrech for reviewing.

**Comment by martinlatrille** [comment]
Thanks for the review guys, I've edited the doc and added a test!

File: apispec/core.py
Comment (martinlatrille) [reply apispec/core.py]: Done!

**Review by sloria — APPROVED** [review]

**Comment by sloria** [comment]
Looks good. Thanks for making those changes.

**Pull request merged**
